### PR TITLE
Litehrnet

### PR DESCRIPTION
**Pull request by jin-s13**
Codes from https://github.com/HRNet/Lite-HRNet.

### CODE REVIEW

**Comment by jin-s13** [comment]
README will be updated after code merge.

File: mmpose/models/backbones/litehrnet.py [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+from .resnet import BasicBlock, Bottleneck
+from .utils import channel_shuffle, load_checkpoint
+
+
+class SpatialWeighting(nn.Module):
+    """Spatial weighting module.
+
+    Args:
+        channels (int): The channels of the module.
+        ratio (int): channel reduction ration.
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:27]: ration

File: mmpose/models/backbones/litehrnet.py [hunk 2]
```diff
@@ -0,0 +1,994 @@
...
+        self.num_branches = num_branches
+
+        self.module_type = module_type
+        self.multiscale_output = multiscale_output
+        self.with_fuse = with_fuse
+        self.norm_cfg = norm_cfg
+        self.conv_cfg = conv_cfg
+        self.with_cp = with_cp
+
+        if self.module_type == 'LITE':
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:558]: make string comparison case insensitive

File: mmpose/models/backbones/litehrnet.py [hunk 3]
```diff
@@ -0,0 +1,994 @@
...
+        return out
+
+
+@BACKBONES.register_module()
+class LiteHRNet(nn.Module):
+    """Lite-HRNet backbone.
+
+    `Lite-HRNet: A Lightweight High-Resolution Network
+    <https://arxiv.org/abs/2104.06403>`__
+
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:741]: "Code adapted from https://github.com/HRNet/Lite-HRNet/blob/hrnet/models/backbones/litehrnet.py"

File: mmpose/models/backbones/litehrnet.py [hunk 4]
```diff
@@ -0,0 +1,994 @@
...
+
+class ConditionalChannelWeighting(nn.Module):
+    """Conditional channel weighting block.
+
+    Args:
+        in_channels (int): The input channels of the block.
+        stride (int): Stride of the 3x3 convolution layer.
+        reduce_ratio (int): channel reduction ration.
+        conv_cfg (dict): Config dict for convolution layer.
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:139]: ration

File: mmpose/models/backbones/litehrnet.py [hunk 5]
```diff
@@ -0,0 +1,994 @@
...
+        for i in range(self.num_stages):
+            num_channels = self.stages_spec['num_channels'][i]
+            num_channels = [num_channels[i] for i in range(len(num_channels))]
+            setattr(
+                self, 'transition{}'.format(i),
+                self._make_transition_layer(num_channels_last, num_channels))
+
+            stage, num_channels_last = self._make_stage(
+                self.stages_spec, i, num_channels, multiscale_output=True)
+            setattr(self, 'stage{}'.format(i), stage)
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:825]: f-str

File: mmpose/models/backbones/litehrnet.py [hunk 6]
```diff
@@ -0,0 +1,994 @@
...
+        self.stages_spec = self.extra['stages_spec']
+
+        num_channels_last = [
+            self.stem.out_channels,
+        ]
+        for i in range(self.num_stages):
+            num_channels = self.stages_spec['num_channels'][i]
+            num_channels = [num_channels[i] for i in range(len(num_channels))]
+            setattr(
+                self, 'transition{}'.format(i),
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:820]: and here

File: mmpose/models/backbones/litehrnet.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/models/backbones/litehrnet.py]: ok.

**Comment by innerlee** [comment]
Looks good

File: mmpose/models/backbones/litehrnet.py [hunk 7]
```diff
@@ -1,8 +1,3 @@
-# ------------------------------------------------------------------------------
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py]: Keep these lines. It is for this whole file. The notice in docstring is for documentation.

File: mmpose/models/backbones/litehrnet.py (hunk 5 above)
Comment (innerlee) [reply mmpose/models/backbones/litehrnet.py]: Search ".format" for old-style interplations

File: mmpose/models/backbones/litehrnet.py [hunk 8]
```diff
@@ -0,0 +1,997 @@
...
+
+    Args:
+        channels (int): The channels of the module.
+        ratio (int): channel reduction ratio.
+        conv_cfg (dict): Config dict for convolution layer.
+            Default: None, which means using conv2d.
+        norm_cfg (dict): Config dict for normalization layer.
+            Default: None.
+        act_cfg (dict): Config dict for activation layer.
+            Default: dict(type='ReLU'), dict(type='Sigmoid'))).
```
Comment (innerlee) [inline mmpose/models/backbones/litehrnet.py:33]: extra `)`

Comment (innerlee) [reply mmpose/models/backbones/litehrnet.py:33]: hmm, actual its wrong doc

**Review by innerlee — APPROVED** [review]

File: mmpose/models/backbones/litehrnet.py (hunk 8 above)
Comment (jin-s13) [reply mmpose/models/backbones/litehrnet.py]: fixed.

File: mmpose/models/backbones/litehrnet.py [hunk 9]
```diff
@@ -0,0 +1,999 @@
...
+import torch.nn.functional as F
+import torch.utils.checkpoint as cp
+from mmcv.cnn import (ConvModule, DepthwiseSeparableConvModule,
+                      build_conv_layer, build_norm_layer, constant_init,
+                      normal_init)
+from torch.nn.modules.batchnorm import _BatchNorm
+
+from mmpose.utils import get_root_logger
+from ..builder import BACKBONES
+from .resnet import BasicBlock, Bottleneck
```
Comment (Junjun2016) [inline mmpose/models/backbones/litehrnet.py:18]: We can remove this line since it is not used.

File: mmpose/models/backbones/litehrnet.py [hunk 10]
```diff
@@ -0,0 +1,999 @@
...
+                    normal_init(m, std=0.001)
+                elif isinstance(m, (_BatchNorm, nn.GroupNorm)):
+                    constant_init(m, 1)
+
+            if self.zero_init_residual:
+                for m in self.modules():
+                    if isinstance(m, Bottleneck):
+                        constant_init(m.norm3, 0)
+                    elif isinstance(m, BasicBlock):
+                        constant_init(m.norm2, 0)
```
Comment (Junjun2016) [inline mmpose/models/backbones/litehrnet.py:965]: We can remove these lines.

File: mmpose/models/backbones/litehrnet.py [hunk 11]
```diff
@@ -0,0 +1,999 @@
...
+        in_channels (int): Number of input image channels. Default: 3.
+        conv_cfg (dict): dictionary to construct and config conv layer.
+        norm_cfg (dict): dictionary to construct and config norm layer.
+        norm_eval (bool): Whether to set norm layers to eval mode, namely,
+            freeze running stats (mean and var). Note: Effect on Batch Norm
+            and its variants only. Default: False
+        with_cp (bool): Use checkpoint or not. Using checkpoint will save some
+            memory while slowing down the training speed.
+        zero_init_residual (bool): whether to use zero init for last norm layer
+            in resblocks to let them behave as identity.
```
Comment (Junjun2016) [inline mmpose/models/backbones/litehrnet.py:758]: Not used.

File: mmpose/models/backbones/litehrnet.py [hunk 12]
```diff
@@ -0,0 +1,999 @@
...
+    def forward(self, x):
+        """Forward function."""
+        x = self.stem(x)
+
+        y_list = [x]
+        for i in range(self.num_stages):
+            x_list = []
+            transition = getattr(self, f'transition{i}')
+            for j in range(self.stages_spec['num_branches'][i]):
+                if transition[j]:
```
Comment (Junjun2016) [inline mmpose/models/backbones/litehrnet.py:978]:
```suggestion
                if transition[j] is not None:
```
```suggestion
                if transition[j]:
```

**Comment by ly015** [comment]
@jin-s13 Could you please update the paper links in .md files according to #778 ?

**Pull request merged**